### DISCUSSION TITLE
Met flush error after on conflict do nothing.

### DISCUSSION THREAD

**Discussion by fockez** [sqlalchemy/sqlalchemy#5856] [Usage Questions] [answered]
**Describe your question**
I intend to insert a lot of rows with unique hash values, and want to skip duplicated records. Using "OR IGNORE" with sqlite3 works fine, but when using "ON CONFLICT DO NOTHING" with pgsql, the session.add passed, but failed at flush, showing flush error like below
`sqlalchemy.orm.exc.FlushError: Instance <tracklets at 0x195ec22b580> has a NULL identity key.  If this is an auto-generated value, check that the database table allows generation of new primary key values, and that the mapped Column object is configured to expect these generated values.  Ensure also that this flush table allows generation of new primary key values, and that the mapped Column object is configured to expect these generated values.  Ensure also that this flush() is not occurring at an inappropriate time, such as within a load() event.`
What should I do and is there a universal method for cross platform insertion with ignore errors, such as an argument of add.

**Versions**
 - OS: Windows 10 x64
 - Python: 3.8.3
 - SQLAlchemy: 1.3.20
 - Database: Postgresql 12
 - DBAPI: psycopg 2.8.6

**Have a nice day!**

**Comment by zzzeek** [comment, marked as answer]
hey there -

the error you are getting means you have an object like this (assume "id' is a primary key attribute):


    obj = MyObject(id=None)

then you try to add it to a session and flush it.  The INSERT statement used by the ORM then did not produce any auto-generated primary key value, so MyObject() can't be placed into the identity map.

The "on conflict" construct supported by the PostgreSQL dialect at https://docs.sqlalchemy.org/en/13/dialects/postgresql.html#insert-on-conflict-upsert is not integrated with the ORM so has nothing to do with this issue, unless you are attempting to somehow integrate the construct using an event of some kind.    

hope this helps!   feel free to add detail as there's no reason given for you getting this error message.
